File: python/open_guardrail/guards/webhook_validate.py
"""Validate webhook requests for integrity."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_HMAC_RE = re.compile(
    r"^[a-fA-F0-9]{64}$|^sha256=[a-fA-F0-9]{64}$"
)
_TS_RE = re.compile(r"^\d{10,13}$")


class _WebhookValidate:
    def __init__(
        self,
        *,
        action: str = "block",
        max_payload_size: int = 1_048_576,
    ) -> None:
        self.name = "webhook-validate"
        self.action = action
        self.max_payload_size = max_payload_size

    def check(
        self,
        text: str,
        stage: str = "input",
        *,
        signature: Optional[str] = None,
        timestamp: Optional[str] = None,
        headers: Optional[dict] = None,
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        if len(text.encode("utf-8")) > self.max_payload_size:
            issues.append("payload_size_exceeded")

        if signature is None:
            issues.append("missing_signature")
        elif not _HMAC_RE.match(signature):
            issues.append("invalid_signature_format")

        if timestamp is None:
            issues.append("missing_timestamp")
        elif not _TS_RE.match(timestamp):
            issues.append("invalid_timestamp_format")

        if headers is not None:
            required = ["content-type"]
            for h in required:
                lower_keys = [
                    k.lower() for k in headers.keys()
                ]
                if h not in lower_keys:
                    issues.append(f"missing_header_{h}")

        triggered = len(issues) > 0
        score = min(len(issues) / 4, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="webhook-validate",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Webhook validation failed"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": issues,
                    "reason": (
                        "Webhook request failed"
                        " integrity checks"
                    ),
                }
                if triggered
                else None
            ),
        )


def webhook_validate(
    *,
    action: str = "block",
    max_payload_size: int = 1_048_576,
) -> _WebhookValidate:
    return _WebhookValidate(
        action=action,
        max_payload_size=max_payload_size,
    )

File: packages/guards/src/webhook-validate.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface WebhookValidateOptions {
  action: 'block' | 'warn';
  maxPayloadSize?: number;
}

const HMAC_PATTERN = /(?:x-hub-signature|x-signature|x-webhook-signature|hmac|sha256=)/i;
const TIMESTAMP_PATTERN = /(?:x-timestamp|x-request-timestamp|t=\d{10,13})/i;
const STALE_TIMESTAMP = /t=(\d{10,13})/;
const FIVE_MINUTES_MS = 5 * 60 * 1000;

export function webhookValidate(options: WebhookValidateOptions): Guard {
  const maxPayloadSize = options.maxPayloadSize ?? 1_048_576;

  return {
    name: 'webhook-validate',
    version: '0.1.0',
    description: 'Validates webhook payloads for HMAC signatures, timestamps, and size limits',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      if (text.length > maxPayloadSize) {
        issues.push(`payload_size_exceeded: ${text.length} > ${maxPayloadSize}`);
      }

      if (!HMAC_PATTERN.test(text)) {
        issues.push('missing_hmac_signature');
      }

      if (!TIMESTAMP_PATTERN.test(text)) {
        issues.push('missing_timestamp');
      }

      const tsMatch = text.match(STALE_TIMESTAMP);
      if (tsMatch) {
        const ts = Number(tsMatch[1]);
        const tsMs = ts < 1e12 ? ts * 1000 : ts;
        if (Math.abs(Date.now() - tsMs) > FIVE_MINUTES_MS) {
          issues.push('stale_timestamp');
        }
      }

      const triggered = issues.length > 0;

      return {
        guardName: 'webhook-validate',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(issues.length / 4, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
